rust/lance/src/dataset/optimize.rs

```diff
@@ -183,6 +208,95 @@ impl CompactionOptions {
     }
 }
 
+/// Determine if page-level binary copy can safely merge the provided fragments.
+///
+/// Preconditions checked in order:
+/// - Feature flag `enable_binary_copy` is enabled
+/// - Dataset storage format is non-legacy
+/// - Fragment list is non-empty
+/// - All data files share identical Lance file versions
+/// - No fragment has a deletion file
+///   TODO need to support schema evolution case like add column and drop column
+/// - All data files share identical schema mappings (`fields`, `column_indices`)
+fn can_use_binary_copy(
+    dataset: &Dataset,
+    options: &CompactionOptions,
+    fragments: &[Fragment],
+) -> bool {
+    use lance_file::version::LanceFileVersion;
+    if !options.enable_binary_copy {
+        return false;
+    }
+
+    // not support blob column for now
+    let has_blob_columns = dataset
+        .schema()
+        .fields_pre_order()
+        .any(|field| field.is_blob());
+    if has_blob_columns {
+        return false;
+    }
+
+    // Check dataset storage version is supported
+    // Binary copy is not supported for legacy Lance file format
+    let storage_ok = dataset
+        .manifest
+        .data_storage_format
+        .lance_file_version()
+        .map(|v| !matches!(v.resolve(), LanceFileVersion::Legacy))
+        .unwrap_or(false);
+    if !storage_ok {
+        return false;
+    }
+
+    if fragments.is_empty() {
+        return false;
+    }
+
+    // Establish version baseline from first data file
+    let first_data_file_version = LanceFileVersion::try_from_major_minor(
+        fragments[0].files[0].file_major_version,
+        fragments[0].files[0].file_minor_version,
+    )
+    .map(|v| v.resolve())
+    .unwrap();
```

Comment:
I think you can just use `dataset.manifest.data_storage_format` like you check above.